azkaban-web-server/src/main/java/cloudflow/services/FlowServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+    if (!lastFlowVersion.isPresent()) {
+      throw new CloudFlowNotFoundException("Latest flow version of flow: " + flow.getId() + "does"
+          + "not exist");
+    }
+    flow.setLastVersion(lastFlowVersion.get());
+  }
+
+
+  @Override
+  public List<FlowResponse> getAllFlows(User user, Map<String, String[]> queryParamMap) {
+    Optional<String> projectId = Optional.empty();
+    Optional<String> projectVersion = Optional.empty();
+    Optional<String> projectName = Optional.empty();
+
+    for (Map.Entry<String, String[]> entry : queryParamMap.entrySet()) {
+      String key = entry.getKey();
+      String[] value = entry.getValue();
+      if (key.equals(PROJECT_ID_PARAM)) {
+        projectId = Optional.of(extractArgumentValue(key, value));
+      } else if (key.equals(PROJECT_VERSION_PARAM)) {
+        projectVersion = Optional.of(extractArgumentValue(key, value));
+      } else if (key.equals(PROJECT_NAME_PARAM)) {
+        projectName = Optional.of(extractArgumentValue(key, value));
+      }
+    }
+
+    /* Both projectId and projectName cannot be passed as query params */
+    if (projectId.isPresent() && projectName.isPresent()) {
+      throw new CloudFlowException("Filtering by both project id and project name is not "
+          + "supported");
+    }
+
+    /* projectVersion cannot be passed as Query param without projectId or projectName */
+    if (!(projectId.isPresent() || projectName.isPresent()) && projectVersion.isPresent()) {
+      throw new CloudFlowException("Project id or project name must be passed along with project "
+          + "version");
+    }
+
+    List<Project> projects = new ArrayList<Project>();
+    /* If projectId or projectName is present, get the project based on query params */
+    if (projectId.isPresent() || projectName.isPresent()) {
+      Optional<Project> project = projectDao
+          .getProjectByParams(projectId, projectName, projectVersion);
+      /* Throw exception if no project was found based on the query params */
+      if (!project.isPresent()) {
+        String errorMsg = String.format("Project record doesn't exist for query params passed");
+        log.error(errorMsg);
+        throw new CloudFlowNotFoundException(errorMsg);
+      }
+      projects.add(project.get());
+    } else {
+      /* No query params passed, get all projects for the user */
+      projects.addAll(projectDao.getAll(user));
+    }
+
+    List<FlowResponse> flows = new ArrayList<FlowResponse>();
+
+    /* If projectVersion is not set, get the flows for the latest project version */
+    String projectVersionStr = "-1";
+    if (projectVersion.isPresent()) {
+      projectVersionStr = projectVersion.get();
+    }
+
+    /* For each project, get all flows and append them to a list */
+    for (Project project : projects) {
+      /* Get all flows for projectId and projectVersion combination */
+      flows.addAll(flowDao.getAllFlows(project.getId(), projectVersionStr));
```

Comment:
Technically all flows for all projects (even with filters) can be fetched by a single SQL query.  This will result in one query for projects and one for flows in each project. 
Any reason for preferring one approach over the other ?

Reply:
Will look into this and talk to rest of the folks about how to restructure the DAOs for joins.